### PR TITLE
pmi+hydra: refactor PMI_process_mapping

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
### rankmap
Refactor to allow arbitrary process mapping and support a general syntax for PMI_process_mapping.

[skip warnings]
```
/* rankmap string format (used in PMI_process_mapping) in ABNF:
 *
 * mapping = '(' format ',' (block / block_repeats) *[',' (block / block_repeats)] ')'
 * format = 'vector'
 * block = num / '(' num ',' num ',' num ')'
 * block_repeats = '[' block *[',' block] ']x' num
 * num = 1*DIGIT
 *
 * If a block is in 3-tuple of (X,Y,Z), the meaning is:
 *     X - node id start value
 *     Y - number of nodes with size Z
 *     Z - number of processes assigned to each node
 *
 * If a block is a single number id, it denotes the node id, which is equivallen to
 * (id, 1, 1).
 *
 * A block_repeat repeats the bracketed block list by num of times.
 */
```
### unit testing
Add a unit testing framework. We specify tests in `test/units` and `python test/units/run_unit.py` will do the testing. The python script will construct a single source program that contains the utility function to be tested and run a list of specified tests against the program.

We can run the unit tests without `configure` or `make`.
```
$ python test/units/run_unit.py
Test - mpl_rankmap_str_to_array
  1:
    ./t 4 '(vector,1,2,3,0)'
    1 2 3 0
    ok
  2:
    ./t 12 '(vector,(0,3,4))'
    0 0 0 0 1 1 1 1 2 2 2 2
    ok
  3:                                                                                                                                                                                                                              
    ./t 12 '(vector,[(0,3,1)]x4)'
    0 1 2 0 1 2 0 1 2 0 1 2
    ok
  4:
    ./t 12 '(vector,[(0,3,1)]x3,(0,3,1))'
    0 1 2 0 1 2 0 1 2 0 1 2
    ok
  5:
    ./t 12 '(vector,[(0,2,1),2]x3,(0,3,1))'
    0 1 2 0 1 2 0 1 2 0 1 2
    ok
  6:
    ./t 12 '(vector,(1,3,1),(0,4,1))'
    1 2 3 0 1 2 3 0 1 2 3 0
    ok
Test - slurm_list_to_nodes
  1:
    ./t host
    host,
    ok
  2:
    ./t 'host-[00-03]'
    host-00,host-01,host-02,host-03,
    ok
  3:
    ./t 'host-[8-11]'
    host-8,host-9,host-10,host-11,
    ok
  4:
    ./t 'host-[8-11,101]'
    host-8,host-9,host-10,host-11,host-101,
    ok
  5:
    ./t '192.168.0.[100-104]'
    192.168.0.100,192.168.0.101,192.168.0.102,192.168.0.103,192.168.0.104,
    ok
  6:
    ./t 'h0-h2'
    h0,h1,h2,
    ok
  7:
    ./t '[h0-h2]'
    h0,h1,h2,
    ok
  8:
    ./t 'a1h[0-2]'
    a1h0,a1h1,a1h2,
    ok
Test - mpl_rankmap_array_to_str
  1:
    ./t 3 2 1 0
    (vector,3,2,1,0)
    ok
  2:
    ./t 1 2 3 0
    (vector,(1,3,1),0)
    ok
  3:
    ./t 0 0 0 0 1 1 1 1 2 2 2 2
    (vector,(0,3,4))
    ok
  4:
    ./t 0 1 2 0 1 2 0 1 2 0 1 2
    (vector,[(0,3,1)]x4)
    ok
  5:
    ./t 1 2 3 0 1 2 3 0 1 2 3 0
    (vector,[(1,3,1),0]x3)
    ok
```

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
Updated quick warnings tests to match the warning flags used in main configure --
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/warnings/auto

File: src/util/mpir_pmi.c [hunk 1]
```diff
@@ -1188,6 +1188,8 @@ static int build_nodemap_pmi2(int *nodemap, int sz);
 static int build_nodemap_pmix(int *nodemap, int sz);
 #endif
 
+static bool do_normalize_nodemap = true;
```
Comment (raffenet) [inline src/util/mpir_pmi.c:1191]: when would this be false?

File: src/util/mpir_pmi.c [hunk 2]
```diff
@@ -1207,6 +1207,30 @@ static int build_nodemap(int *nodemap, int sz, int *p_max_node_id)
...
 
+    /* node ids from process manager may not start from 0 or has gaps.
+     * Normalize it since most of the code assume a contiguous node id range */
+    int max_id = -1;
+    for (int i = 0; i < sz, i++) {
+        if (max_id < nodemap[i]) {
+            max_id = nodemap[i];
+        }
+    }
+    int *nodeids = MPL_malloc((max_id + 1) * sizeof(int), MPL_MEM_OTHER);
```
[the remark this answers is not in the document]
Comment (raffenet) [reply src/util/mpir_pmi.c]: Nevermind, I see in a later commit the free is added.

**Review by raffenet — COMMENTED** [review]
just a couple minor comments

File: src/util/mpir_pmi.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/util/mpir_pmi.c]: Good catch. I added the missing `MPL_free` to the initial commit for completeness.

File: src/util/mpir_pmi.c (hunk 1 above)
Comment (hzhou) [reply src/util/mpir_pmi.c:1197]: I added a TODO comment.

**Comment by hzhou** [comment]
test:mpich/pmi

**Review by raffenet — APPROVED** [review]
Looks good.

**Pull request merged**